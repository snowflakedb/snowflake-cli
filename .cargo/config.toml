# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Rust configuration for MAXIMUM x86_64 compatibility
# Use the most conservative settings possible to avoid "Illegal instruction" errors
# This is for CI x86_64 Linux builds, not cross-compilation

[build]
# Let cargo use the native target but with conservative flags

[target.x86_64-unknown-linux-gnu]
# Use only baseline x86_64 instruction set (SSE/SSE2 only)
# Disable all optional/newer instruction sets for maximum compatibility
rustflags = [
    "-C", "target-cpu=core2",
    "-C", "target-feature=-sse3,-ssse3,-sse4.1,-sse4.2,-popcnt,-avx,-avx2,-fma,-bmi1,-bmi2,-lzcnt,-movbe",
    "-C", "opt-level=2",
    "-C", "lto=false"
]
