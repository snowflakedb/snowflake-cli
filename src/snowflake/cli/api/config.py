# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import logging
import os
import warnings
from contextlib import contextmanager
from dataclasses import asdict, dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional, Union

import tomlkit
from click import ClickException
from snowflake.cli.api.exceptions import (
    ConfigFileTooWidePermissionsError,
    MissingConfiguration,
    UnsupportedConfigSectionTypeError,
)
from snowflake.cli.api.secure_path import SecurePath
from snowflake.cli.api.secure_utils import (
    file_permissions_are_strict,
    windows_get_not_whitelisted_users_with_access,
)
from snowflake.cli.api.utils.types import try_cast_to_bool
from snowflake.connector.compat import IS_WINDOWS
from snowflake.connector.config_manager import CONFIG_MANAGER
from snowflake.connector.constants import CONFIG_FILE, CONNECTIONS_FILE
from snowflake.connector.errors import ConfigSourceError, MissingConfigOptionError
from tomlkit import TOMLDocument, dump
from tomlkit.container import Container
from tomlkit.exceptions import NonExistentKey
from tomlkit.items import Table

log = logging.getLogger(__name__)


class Empty:
    pass


CONNECTIONS_SECTION = "connections"
CLI_SECTION = "cli"
LOGS_SECTION = "logs"
PLUGINS_SECTION = "plugins"

LOGS_SECTION_PATH = [CLI_SECTION, LOGS_SECTION]
PLUGINS_SECTION_PATH = [CLI_SECTION, PLUGINS_SECTION]
PLUGIN_ENABLED_KEY = "enabled"
FEATURE_FLAGS_SECTION_PATH = [CLI_SECTION, "features"]

CONFIG_MANAGER.add_option(
    name=CLI_SECTION,
    parse_str=tomlkit.parse,
    default=dict(),
)


@dataclass
class ConnectionConfig:
    account: Optional[str] = None
    user: Optional[str] = None
    password: Optional[str] = field(default=None, repr=False)
    host: Optional[str] = None
    region: Optional[str] = None
    port: Optional[int] = None
    database: Optional[str] = None
    schema: Optional[str] = None
    warehouse: Optional[str] = None
    role: Optional[str] = None
    authenticator: Optional[str] = None
    private_key_file: Optional[str] = None
    token_file_path: Optional[str] = None

    _other_settings: dict = field(default_factory=lambda: {})

    @classmethod
    def from_dict(cls, config_dict: dict) -> ConnectionConfig:
        known_settings = {}
        other_settings = {}
        for key, value in config_dict.items():
            if key in cls.__dict__:
                known_settings[key] = value
            else:
                other_settings[key] = value
        return cls(**known_settings, _other_settings=other_settings)

    def to_dict_of_known_non_empty_values(self) -> dict:
        return {
            k: v
            for k, v in asdict(self).items()
            if k != "_other_settings" and v is not None
        }

    def _non_empty_other_values(self) -> dict:
        return {k: v for k, v in self._other_settings.items() if v is not None}

    def to_dict_of_all_non_empty_values(self) -> dict:
        return {
            **self.to_dict_of_known_non_empty_values(),
            **self._non_empty_other_values(),
        }


def config_init(config_file: Optional[Path]):
    """
    Initializes the app configuration. Config provided via cli flag takes precedence.
    If config file does not exist we create an empty one.
    """
    from snowflake.cli._app.loggers import create_initial_loggers

    if config_file:
        CONFIG_MANAGER.file_path = config_file
    else:
        _check_default_config_files_permissions()
    if not CONFIG_MANAGER.file_path.exists():
        _initialise_config(CONFIG_MANAGER.file_path)
    _read_config_file()
    create_initial_loggers()


def add_connection_to_proper_file(name: str, connection_config: ConnectionConfig):
    if CONNECTIONS_FILE.exists():
        existing_connections = _read_connections_toml()
        existing_connections.update(
            {name: connection_config.to_dict_of_all_non_empty_values()}
        )
        _update_connections_toml(existing_connections)
        return CONNECTIONS_FILE
    else:
        set_config_value(
            path=[CONNECTIONS_SECTION, name],
            value=connection_config.to_dict_of_all_non_empty_values(),
        )
        return CONFIG_MANAGER.file_path


_DEFAULT_LOGS_CONFIG = {
    "save_logs": True,
    "path": str(CONFIG_MANAGER.file_path.parent / "logs"),
    "level": "info",
}

_DEFAULT_CLI_CONFIG = {LOGS_SECTION: _DEFAULT_LOGS_CONFIG}


@contextmanager
def _config_file():
    _read_config_file()
    conf_file_cache = CONFIG_MANAGER.conf_file_cache
    yield conf_file_cache
    _dump_config(conf_file_cache)


def _read_config_file():
    with warnings.catch_warnings():
        if IS_WINDOWS:
            warnings.filterwarnings(
                action="ignore",
                message="Bad owner or permissions.*",
                module="snowflake.connector.config_manager",
            )

            if not file_permissions_are_strict(CONFIG_MANAGER.file_path):
                users = ", ".join(
                    windows_get_not_whitelisted_users_with_access(
                        CONFIG_MANAGER.file_path
                    )
                )
                warnings.warn(
                    f"Unauthorized users ({users}) have access to configuration file {CONFIG_MANAGER.file_path}.\n"
                    f'Run `icacls "{CONFIG_MANAGER.file_path}" /remove:g <USER_ID>` on those users to restrict permissions.'
                )

        try:
            CONFIG_MANAGER.read_config()
        except ConfigSourceError as exception:
            raise ClickException(
                f"Configuration file seems to be corrupted. {str(exception.__cause__)}"
            )


def _initialise_logs_section():
    with _config_file() as conf_file_cache:
        if conf_file_cache.get(CLI_SECTION) is None:
            conf_file_cache[CLI_SECTION] = _DEFAULT_CLI_CONFIG
        if conf_file_cache[CLI_SECTION].get(LOGS_SECTION) is None:
            conf_file_cache[CLI_SECTION][LOGS_SECTION] = _DEFAULT_LOGS_CONFIG


def set_config_value(path: List[str], value: Any) -> None:
    """Sets value in config.
    For example to set value "val" to key "key" in section [a.b.c], call
    set_config_value(["a", "b", "c", "key"], "val").
    If you want to override a whole section, value should be a dictionary.
    """
    with _config_file() as conf_file_cache:
        current_config_dict = conf_file_cache
        for key in path[:-1]:
            if key not in current_config_dict:
                current_config_dict[key] = {}
            current_config_dict = current_config_dict[key]
        current_config_dict[path[-1]] = value


def get_logs_config() -> dict:
    logs_config = _DEFAULT_LOGS_CONFIG.copy()
    if config_section_exists(*LOGS_SECTION_PATH):
        logs_config.update(**get_config_section(*LOGS_SECTION_PATH))
    return logs_config


def get_plugins_config() -> dict:
    if config_section_exists(*PLUGINS_SECTION_PATH):
        return get_config_section(*PLUGINS_SECTION_PATH)
    else:
        return {}


def connection_exists(connection_name: str) -> bool:
    return config_section_exists(CONNECTIONS_SECTION, connection_name)


def config_section_exists(*path) -> bool:
    try:
        _find_section(*path)
        return True
    except (KeyError, NonExistentKey, MissingConfigOptionError):
        return False


def get_all_connections() -> dict[str, ConnectionConfig]:
    return {
        k: ConnectionConfig.from_dict(connection_dict)
        for k, connection_dict in get_config_section("connections").items()
    }


def get_connection_dict(connection_name: str) -> dict:
    try:
        return get_config_section(CONNECTIONS_SECTION, connection_name)
    except KeyError:
        raise MissingConfiguration(f"Connection {connection_name} is not configured")


def get_default_connection_name() -> str:
    return CONFIG_MANAGER["default_connection_name"]


def get_default_connection_dict() -> dict:
    def_connection_name = get_default_connection_name()
    if not connection_exists(def_connection_name):
        raise MissingConfiguration(
            f"Couldn't find connection for default connection `{def_connection_name}`. "
            f"Specify connection name or configure default connection."
        )
    return get_connection_dict(def_connection_name)


def get_config_section(*path) -> dict:
    section = _find_section(*path)
    if isinstance(section, Container):
        return {s: _merge_section_with_env(section[s], *path, s) for s in section}
    if isinstance(section, dict):
        return _merge_section_with_env(section, *path)
    raise UnsupportedConfigSectionTypeError(type(section))


def get_config_value(*path, key: str, default: Optional[Any] = Empty) -> Any:
    """Looks for given key under nested path in toml file."""
    env_variable = get_env_value(*path, key=key)
    if env_variable:
        return env_variable
    try:
        return get_config_section(*path)[key]
    except (KeyError, NonExistentKey, MissingConfigOptionError):
        if default is not Empty:
            return default
        raise


def get_config_bool_value(*path, key: str, default: Optional[bool]) -> Optional[bool]:
    value = get_config_value(*path, key=key, default=None)

    if value is None:
        return default

    try:
        return try_cast_to_bool(value)
    except ValueError:
        raise ClickException(
            f"Expected boolean value for {'.'.join((*path, key))} option."
        )


def _initialise_config(config_file: Path) -> None:
    config_file = SecurePath(config_file)
    config_file.parent.mkdir(parents=True, exist_ok=True)
    config_file.touch()
    _initialise_logs_section()
    log.info("Created Snowflake configuration file at %s", CONFIG_MANAGER.file_path)


def get_env_variable_name(*path, key: str) -> str:
    return ("_".join(["snowflake", *path, key])).upper()


def get_env_value(*path, key: str) -> str | None:
    return os.environ.get(get_env_variable_name(*path, key=key))


def _find_section(*path) -> TOMLDocument:
    section = CONFIG_MANAGER
    idx = 0
    while idx < len(path):
        section = section[path[idx]]
        idx += 1
    return section


def _merge_section_with_env(section: Union[Table, Any], *path) -> Dict[str, str]:
    if isinstance(section, Table):
        env_variables = _get_envs_for_path(*path)
        section_copy = section.copy()
        section_copy.update(env_variables)
        return section_copy.unwrap()
    # It's a atomic value
    return section


def _get_envs_for_path(*path) -> dict:
    env_variables_prefix = "_".join(["SNOWFLAKE"] + [p.upper() for p in path]) + "_"
    return {
        k.replace(env_variables_prefix, "").lower(): os.environ[k]
        for k in os.environ.keys()
        if k.startswith(env_variables_prefix)
    }


def _dump_config(config_and_connections: Dict):
    config_toml_dict = config_and_connections.copy()

    if CONNECTIONS_FILE.exists():
        # update connections in connections.toml
        # it will add only connections (maybe updated) which were originally read from connections.toml
        # it won't add connections from config.toml
        # because config manager doesn't have connections from config.toml if connections.toml exists
        _update_connections_toml(config_and_connections.get("connections") or {})
        # to config.toml save only connections from config.toml
        connections_to_save_in_config_toml = _read_config_file_toml().get("connections")
        if connections_to_save_in_config_toml:
            config_toml_dict["connections"] = connections_to_save_in_config_toml
        else:
            config_toml_dict.pop("connections", None)

    with SecurePath(CONFIG_MANAGER.file_path).open("w+") as fh:
        dump(config_toml_dict, fh)


def _check_default_config_files_permissions() -> None:
    if not IS_WINDOWS:
        if CONNECTIONS_FILE.exists() and not file_permissions_are_strict(
            CONNECTIONS_FILE
        ):
            raise ConfigFileTooWidePermissionsError(CONNECTIONS_FILE)
        if CONFIG_FILE.exists() and not file_permissions_are_strict(CONFIG_FILE):
            raise ConfigFileTooWidePermissionsError(CONFIG_FILE)


from typing import Literal


def get_feature_flags_section() -> Dict[str, bool | Literal["UNKNOWN"]]:
    if not config_section_exists(*FEATURE_FLAGS_SECTION_PATH):
        return {}

    flags = get_config_section(*FEATURE_FLAGS_SECTION_PATH)

    def _bool_or_unknown(value):
        try:
            return try_cast_to_bool(value)
        except ValueError:
            return "UNKNOWN"

    return {k: _bool_or_unknown(v) for k, v in flags.items()}


def _read_config_file_toml() -> dict:
    return tomlkit.loads(CONFIG_MANAGER.file_path.read_text()).unwrap()


def _read_connections_toml() -> dict:
    return tomlkit.loads(CONNECTIONS_FILE.read_text()).unwrap()


def _update_connections_toml(connections: dict):
    with open(CONNECTIONS_FILE, "w") as f:
        f.write(tomlkit.dumps(connections))
