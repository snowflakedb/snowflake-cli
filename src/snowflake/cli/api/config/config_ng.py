# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Next-generation configuration system (config-ng).

This module will contain the new configuration handling implementation.
It provides a modern, type-safe, and extensible configuration system
that will eventually replace the legacy configuration system.

TODO: Implement the new configuration system with:
- Type-safe configuration schemas using Pydantic or similar
- Better validation and error handling
- Improved environment variable support with type coercion
- Plugin-based configuration providers
- Configuration migration utilities from legacy system
- Support for multiple configuration formats (TOML, YAML, JSON)
- Configuration templates and inheritance
- Better error messages with suggestions
"""

from __future__ import annotations

import logging
from pathlib import Path
from typing import Optional

log = logging.getLogger(__name__)

# TODO: Implement new configuration system
# This is a clean implementation that will help identify what needs to be implemented
# by failing fast when config-ng is enabled but functionality is missing


def config_init(config_file: Optional[Path]) -> None:
    """
    Initialize the next-generation configuration system.

    This is a placeholder for the new configuration initialization.
    When config-ng is enabled, this function should be fully implemented
    without relying on legacy code.

    Args:
        config_file: Optional path to configuration file

    TODO: Implement new features:
    - Configuration schema validation
    - Better error handling and user feedback
    - Configuration migration from legacy format
    - Support for configuration inheritance
    - Environment-specific configuration overlays
    """
    log.info("Using next-generation configuration system (config-ng)")
    # TODO: Implement new config initialization
    raise NotImplementedError(
        "config_init is not implemented in config-ng. "
        "This helps identify what needs to be implemented for the new system."
    )


# TODO: Add all other configuration functions and constants that are needed
# Each missing function will cause an ImportError when config-ng is enabled,
# helping identify what needs to be implemented

# Example of what needs to be implemented:
# - ConnectionConfig class
# - CONFIG_MANAGER equivalent
# - All configuration constants (CLI_SECTION, CONNECTIONS_SECTION, etc.)
# - All configuration functions (set_config_value, get_config_value, etc.)

# For now, we'll let ImportError guide us to what needs to be implemented
