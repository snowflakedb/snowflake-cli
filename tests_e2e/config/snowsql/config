[connections]

accountname = "default_connection_account"
username = "default_connection_user"
host = "localhost"
databasename = "default_connection_database"
schemaname = "public"
warehousename = "default_connection_warehouse"
rolename = "accountadmin"


[connections.snowsql1]
accountname = "a1"
username = "u1"
host = "h1"
databasename = "d1"
schemaname = "public"
warehousename = "w1"
rolename = "r1"

[connections.snowsql2]
accountname = "a2"
username = "u2"
host = "h2"
databasename = "d2"
schemaname = "public"
warehousename = "w2"
rolename = "r2"
port = 1234

[connections.example]
accountname = accountname
username = username


[variables]
example_variable=27


[options]
auto_completion = True
log_file = /tmp/snowsql.log
log_level = DEBUG
timing = True
output_format = psql
key_bindings = emacs
repository_base_url = https://sfc-repo.snowflakecomputing.com/snowsql
