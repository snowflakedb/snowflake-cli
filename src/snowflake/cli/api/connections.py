# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import asyncio
import logging
import re
import warnings
from dataclasses import asdict, dataclass, field, fields, replace
from pathlib import Path
from typing import Optional

from snowflake.cli.api.config import get_connection_dict, get_default_connection_name
from snowflake.cli.api.exceptions import InvalidSchemaError
from snowflake.connector import SnowflakeConnection
from snowflake.connector.compat import IS_WINDOWS

logger = logging.getLogger(__name__)

schema_pattern = re.compile(r".+\..+")


@dataclass
class ConnectionContext:
    # FIXME: can reduce duplication using config.ConnectionConfig
    connection_name: Optional[str] = None
    host: Optional[str] = None
    port: Optional[int] = None
    account: Optional[str] = None
    database: Optional[str] = None
    role: Optional[str] = None
    schema: Optional[str] = None
    user: Optional[str] = None
    password: Optional[str] = field(default=None, repr=False)
    authenticator: Optional[str] = None
    private_key_file: Optional[str] = None
    warehouse: Optional[str] = None
    mfa_passcode: Optional[str] = None
    enable_diag: Optional[bool] = False
    diag_log_path: Optional[Path] = None
    diag_allowlist_path: Optional[Path] = None
    temporary_connection: bool = False
    session_token: Optional[str] = None
    master_token: Optional[str] = None
    token_file_path: Optional[Path] = None

    VALIDATED_FIELD_NAMES = ["schema"]

    def present_values_as_dict(self) -> dict:
        """Dictionary representation of this ConnectionContext for values that are not None"""
        return {k: v for (k, v) in asdict(self).items() if v is not None}

    def clone(self) -> ConnectionContext:
        return replace(self)

    def update(self, **updates):
        """
        Given a dictionary of property (key, value) mappings, update properties
        of this context object with equivalent names to the keys.

        Raises KeyError if a non-property is specified as a key.
        """
        field_map = {field.name: field for field in fields(self)}
        for key, value in updates.items():
            # ensure key represents a property
            if key not in field_map:
                raise KeyError(f"{key} is not a field of {self.__class__.__name__}")
            setattr(self, key, value)

    def merge_with_config(self, **updates) -> ConnectionContext:
        """
        Updates missing fields from the config, but does not overwrite existing.
        """
        field_map = {field.name for field in fields(self)}
        for key, value in updates.items():
            if key in field_map and getattr(self, key) is None:
                setattr(self, key, value)

        return self

    def update_from_config(self) -> ConnectionContext:
        connection_config = get_connection_dict(connection_name=self.connection_name)
        if "private_key_path" in connection_config:
            connection_config["private_key_file"] = connection_config[
                "private_key_path"
            ]
            del connection_config["private_key_path"]

        self.merge_with_config(**connection_config)
        return self

    def __repr__(self) -> str:
        """Minimal repr where None values have their keys omitted."""
        items = [f"{k}={repr(v)}" for (k, v) in self.present_values_as_dict().items()]
        return f"{self.__class__.__name__}({', '.join(items)})"

    def __setattr__(self, prop, val):
        """Runs registered validators before setting fields."""
        if prop in self.VALIDATED_FIELD_NAMES:
            validate = getattr(self, f"validate_{prop}")
            validate(val)
        super().__setattr__(prop, val)

    def validate_schema(self, value: Optional[str]):
        if (
            value
            and not (value.startswith('"') and value.endswith('"'))
            # if schema is fully qualified name (db.schema)
            and schema_pattern.match(value)
        ):
            raise InvalidSchemaError(value)

    def validate_and_complete(self):
        """
        Ensure we can create a connection from this context.
        """
        if not self.temporary_connection and not self.connection_name:
            self.connection_name = get_default_connection_name()

    def build_connection(self):
        from snowflake.cli._app.snow_connector import connect_to_snowflake

        # Ignore warnings about bad owner or permissions on Windows
        # Telemetry omit our warning filter from config.py
        if IS_WINDOWS:
            warnings.filterwarnings(
                action="ignore",
                message="Bad owner or permissions.*",
                module="snowflake.connector.config_manager",
            )

        return connect_to_snowflake(**self.present_values_as_dict())


class OpenConnectionCache:
    """
    A connection cache that transparently manages SnowflakeConnection objects
    and is keyed by ConnectionContext objects, e.g. cache[ctx].execute_string(...).
    Connections are automatically closed after CONNECTION_CLEANUP_SEC, but
    are guaranteed to be open (if config is valid) when returned by the cache.
    """

    connections: dict[str, SnowflakeConnection]
    cleanup_futures: dict[str, asyncio.TimerHandle]

    CONNECTION_CLEANUP_SEC: float = 10.0 * 60
    """Connections are closed this many seconds after the last time they are accessed."""

    def __init__(self):
        self.connections = {}
        self.cleanup_futures = {}

    def __getitem__(self, ctx):
        if not isinstance(ctx, ConnectionContext):
            raise ValueError(
                f"Expected key to be ConnectionContext but got {repr(ctx)}"
            )
        key = repr(ctx)
        if not self._has_open_connection(key):
            self._insert(key, ctx)
        self._touch(key)
        return self.connections[key]

    def clear(self):
        """Closes all connections and resets the cache to its initial state."""
        connection_keys = list(self.connections.keys())
        for key in connection_keys:
            self._cleanup(key)

        # if any orphaned futures still exist, clean them up too
        for future in self.cleanup_futures.values():
            future.cancel()
        self.cleanup_futures.clear()

    def _has_open_connection(self, key: str):
        return key in self.connections

    def _insert(self, key: str, ctx: ConnectionContext):
        try:
            # N.B. build_connection ultimately calls connect_to_snowflake, which
            # interpolates in connection dicts (from config) and environment variables.
            # This means that we could return a stale (incorrect) connection for the
            # given ConnectionContext if get_env_value or get_connection_dict would
            # have returned different values (i.e. env / config have changed).
            self.connections[key] = ctx.build_connection()
        except Exception:
            logger.debug(
                "ConnectionCache: failed to connect using %s; not caching.", key
            )
            raise

    def _cancel_cleanup_future_if_exists(self, key: str):
        if key in self.cleanup_futures:
            self.cleanup_futures.pop(key).cancel()

    def _touch(self, key: str):
        """
        Extend the lifetime of the cached connection at the given key.
        """
        loop = None
        try:
            loop = asyncio.get_event_loop()
        except RuntimeError:
            # Python 3.11+ will throw when no event loop;
            # Python 3.10 will issue a DeprecationWarning and return None
            pass

        if not loop:
            logger.debug(
                "ConnectionCache: no event loop; connections will close at exit."
            )
            return

        self._cancel_cleanup_future_if_exists(key)
        self.cleanup_futures[key] = loop.call_later(
            self.CONNECTION_CLEANUP_SEC, lambda: self._cleanup(key)
        )

    def _cleanup(self, key: str):
        """Closes the cached connection at the given key."""
        if key not in self.connections:
            logger.debug("Cleaning up connection %s, but not found in cache!", key)

        # doesn't cancel in-flight async queries
        self._cancel_cleanup_future_if_exists(key)
        self.connections.pop(key).close()
