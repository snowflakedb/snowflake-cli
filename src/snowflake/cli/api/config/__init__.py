# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Configuration package with support for legacy and next-generation config systems.

This package provides a feature flag mechanism to switch between the legacy
configuration system and the new config-ng system using the environment variable
SNOWFLAKE_CLI_CONFIG_NG.

Usage:
    # Enable new config system
    export SNOWFLAKE_CLI_CONFIG_NG=1

    # Use legacy config system (default)
    unset SNOWFLAKE_CLI_CONFIG_NG

Environment Variables:
    SNOWFLAKE_CLI_CONFIG_NG: Set to '1', 'true', 'yes', or 'on' to enable config-ng
"""

from __future__ import annotations

import os
from typing import Literal

# Environment variable to enable the new config-ng system
CONFIG_NG_ENV_VAR: Literal["SNOWFLAKE_CLI_CONFIG_NG"] = "SNOWFLAKE_CLI_CONFIG_NG"


def _is_config_ng_enabled() -> bool:
    """
    Check if the new config-ng system should be used.

    Returns:
        True if config-ng should be used, False for legacy system
    """
    return os.environ.get(CONFIG_NG_ENV_VAR, "").lower() in ("1", "true", "yes", "on")


# Conditionally import from legacy or config-ng based on environment variable
if _is_config_ng_enabled():
    # Import ONLY from the new config-ng system
    # This will fail if config-ng is not fully implemented, which is intentional
    # to help identify what needs to be implemented
    try:
        from .config_ng import *  # noqa: F403, F401
    except ImportError as e:
        # If config-ng is not fully implemented, fail fast with a clear error
        raise ImportError(
            f"Config-ng is enabled via {CONFIG_NG_ENV_VAR} but is not fully implemented. "
            f"Missing implementation: {e}. "
            f"Either implement the missing functionality in config_ng.py or "
            f"disable config-ng by unsetting {CONFIG_NG_ENV_VAR}."
        ) from e
else:
    # Use the legacy configuration system
    from .legacy import *  # noqa: F403, F401

# Export the feature flag utilities and factory method for external use
__all__ = ["_is_config_ng_enabled", "CONFIG_NG_ENV_VAR", "get_config_manager"]
