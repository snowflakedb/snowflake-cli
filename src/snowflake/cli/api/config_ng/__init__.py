# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Enhanced Configuration System - Next Generation (NG)

This package implements a simple, extensible configuration system with:
- Two-phase resolution: file sources use connection-level replacement,
  overlay sources (env/CLI) use field-level merging
- List-order precedence (explicit ordering in source list)
- Migration support (SnowCLI and SnowSQL compatibility)
- Complete resolution history tracking
- Read-only, immutable configuration sources
"""

from snowflake.cli.api.config_ng.core import (
    ConfigValue,
    ResolutionEntry,
    ResolutionHistory,
    SourceType,
    ValueSource,
)
from snowflake.cli.api.config_ng.presentation import ResolutionPresenter
from snowflake.cli.api.config_ng.resolution_logger import (
    check_value_source,
    explain_configuration,
    export_resolution_history,
    format_summary_for_display,
    get_resolution_summary,
    get_resolver,
    is_resolution_logging_available,
    show_all_resolution_chains,
    show_resolution_chain,
)
from snowflake.cli.api.config_ng.resolver import (
    ConfigurationResolver,
    ResolutionHistoryTracker,
)
from snowflake.cli.api.config_ng.sources import (
    CliConfigFile,
    CliEnvironment,
    CliParameters,
    ConnectionsConfigFile,
    ConnectionSpecificEnvironment,
    SnowSQLConfigFile,
    SnowSQLEnvironment,
    SnowSQLSection,
    get_merged_variables,
)

__all__ = [
    "check_value_source",
    "CliConfigFile",
    "CliEnvironment",
    "CliParameters",
    "ConfigurationResolver",
    "ConfigValue",
    "ConnectionsConfigFile",
    "ConnectionSpecificEnvironment",
    "explain_configuration",
    "export_resolution_history",
    "format_summary_for_display",
    "get_merged_variables",
    "get_resolution_summary",
    "get_resolver",
    "is_resolution_logging_available",
    "ResolutionEntry",
    "ResolutionHistory",
    "ResolutionHistoryTracker",
    "ResolutionPresenter",
    "show_all_resolution_chains",
    "show_resolution_chain",
    "SnowSQLConfigFile",
    "SnowSQLEnvironment",
    "SnowSQLSection",
    "SourceType",
    "ValueSource",
]
