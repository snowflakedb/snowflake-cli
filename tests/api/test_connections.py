# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from dataclasses import asdict
from unittest import mock

import pytest
from snowflake.cli.api.connections import ConnectionContext, OpenConnectionCache


@pytest.fixture
def local_connection_cache():
    cache = OpenConnectionCache()
    yield cache
    cache.clear()


@pytest.mark.parametrize(
    "args",
    [
        {},
        {"connection_name": "myconn"},
        {"temporary_connection": True, "account": "myacct", "user": "myuser"},
    ],
)
def test_stable_connection_context_repr(args: dict, snapshot):
    ctx = ConnectionContext()
    ctx.update(**args)
    ctx.validate_and_complete()
    assert repr(ctx) == snapshot


def test_clone_connection_context():
    """
    Tests that the clone() method is working properly.
    """

    keys = (
        "connection_name",
        "account",
        "database",
        "role",
        "schema",
        "user",
        "password",
        "authenticator",
        "warehouse",
        "session_token",
        "master_token",
    )

    old_ctx = ConnectionContext()
    for key in keys:
        setattr(old_ctx, key, "value")

    new_ctx = old_ctx.clone()
    assert asdict(new_ctx) == asdict(old_ctx)

    for key in keys:
        setattr(new_ctx, key, "updated_values_should_not_appear_in_old_ctx")
        assert getattr(old_ctx, key) == "value"


@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._app.snow_connector.command_info")
def test_connection_cache_caches(
    mock_command_info, mock_connect, local_connection_cache, test_snowcli_config
):
    mock_command_info.return_value = "application"

    from snowflake.cli.api.config import config_init

    config_init(test_snowcli_config)

    ctx = ConnectionContext(connection_name="default")

    local_connection_cache[ctx]
    local_connection_cache[ctx]
    local_connection_cache[ctx]

    mock_connect.assert_called_once_with(
        application=mock_command_info.return_value,
        database="db_for_test",
        schema="test_public",
        role="test_role",
        warehouse="xs",
        password="dummy_password",
        application_name="snowcli",
        using_session_keep_alive=True,
    )
