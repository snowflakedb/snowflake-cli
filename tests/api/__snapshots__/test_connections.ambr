# serializer version: 1
# name: test_stable_connection_context_repr[args0]
  "ConnectionContext(connection_name='default', enable_diag=False, temporary_connection=False)"
# ---
# name: test_stable_connection_context_repr[args1]
  "ConnectionContext(connection_name='myconn', enable_diag=False, temporary_connection=False)"
# ---
# name: test_stable_connection_context_repr[args2]
  "ConnectionContext(account='myacct', user='myuser', enable_diag=False, temporary_connection=True)"
# ---
